third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+        &self,
+        state: &mut LifetimeState,
+        code_offset: CodeOffset,
+        id: AttrId,
+        struct_: QualifiedInstId<StructId>,
+        field_offs: &usize,
+        dest: TempIndex,
+        src: TempIndex,
+        alive: &LiveVarInfoAtCodeOffset,
+    ) {
+        let label = state.make_local(src, code_offset, 0);
+        let descendant = state.replace_local(dest, code_offset, 1);
+        let loc = self.loc(id);
+        let is_mut = self.ty(dest).is_mutable_reference();
+        let struct_env = self
+            .target
+            .global_env()
+            .get_struct(struct_.to_qualified_id());
+        let field_id = struct_env.get_field_by_offset(*field_offs).get_id();
+        self.check_and_add_edge(
+            state,
+            label,
+            BorrowEdge::new(
+                BorrowEdgeKind::BorrowField(field_id),
+                is_mut,
+                Some(loc),
+                descendant,
+            ),
+            alive,
+        )
+    }
+
+    /// Process a function call. For now we implement standard Move semantics, where every
+    /// output reference is a descendant of all input references. Here would be the point where to
+    //  evaluate lifetime modifiers in future language versions.
+    fn call_operation(
+        &self,
+        state: &mut LifetimeState,
+        code_offset: CodeOffset,
+        id: AttrId,
+        oper: Operation,
+        dests: &[TempIndex],
+        srcs: &[TempIndex],
+        alive: &LiveVarInfoAtCodeOffset,
+    ) {
+        // Check validness of transferring arguments into function.
+        let mut src_check_failed = BTreeSet::new();
+        for src in srcs {
+            if !self.check_read_local(state, id, *src, ReadMode::Argument, alive) {
+                src_check_failed.insert(*src);
+            }
+        }
+        // Next check whether we can assign to the destinations.
+        for dest in dests {
+            self.check_write_local(state, id, *dest, alive)
+        }
+        // Now draw edges from all reference destinations to all reference sources
+        let dest_labels = dests
+            .iter()
+            .filter(|d| self.ty(**d).is_reference())
+            .enumerate()
+            .map(|(i, t)| (*t, state.replace_local(*t, code_offset, i as u8)))
+            .collect::<BTreeMap<_, _>>();
+        let src_qualifier_offset = dest_labels.len();
+        let loc = self.loc(id);
+        for dest in dests {
+            let dest_ty = self.ty(*dest);
+            if dest_ty.is_reference() {
+                for (i, src) in srcs.iter().enumerate() {
+                    // Only check the edge if src check succeeded, otherwise we get confusing
+                    // double errors on the same location.
+                    if src_check_failed.contains(src) {
+                        continue;
+                    }
+                    let src_ty = self.ty(*src);
+                    if src_ty.is_reference() {
```

Comment:
nit: would it be better to do a filter on `srcs` to keep only the references, instead of doing that filter once per `dest`?

Reply:
Bit ugly to do this because I need the src type to filter, but I also need it in the body.